In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import clean # our .py file containing the cleaning functions

In [2]:
reviews = pd.read_csv('https://query.data.world/s/55cb4g2ccy2sbat45jzrwmfjfkp2d5?dws=00000')
reviews.review_time = pd.to_datetime(reviews.review_time,unit = 's')

In [3]:
reviews

brewery_id                 brewery_name         review_time  \
0             10325              Vecchio Birraio 2009-02-16 20:57:03   
1             10325              Vecchio Birraio 2009-03-01 13:44:57   
2             10325              Vecchio Birraio 2009-03-01 14:10:04   
3             10325              Vecchio Birraio 2009-02-15 19:12:25   
4              1075      Caldera Brewing Company 2010-12-30 18:53:26   
...             ...                          ...                 ...   
1586609       14359  The Defiant Brewing Company 2006-11-05 00:01:32   
1586610       14359  The Defiant Brewing Company 2006-10-17 01:29:26   
1586611       14359  The Defiant Brewing Company 2006-10-13 01:21:53   
1586612       14359  The Defiant Brewing Company 2006-10-05 04:37:24   
1586613       14359  The Defiant Brewing Company 2006-10-04 23:41:59   

         review_overall  review_aroma  review_appearance review_profilename  \
0                   1.5           2.0                2.5            stcules   
1                   3.0           2.5                3.0            stcules   
2                   3.0           2.5                3.0            stcules   
3                   3.0           3.0                3.5            stcules   
4                   4.0           4.5                4.0     johnmichaelsen   
...                 ...           ...                ...                ...   
1586609             5.0           4.0                3.5         maddogruss   
1586610             4.0           5.0                2.5          yelterdow   
1586611             4.5           3.5                3.0           TongoRad   
1586612             4.0           4.5                4.5           dherling   
1586613             5.0           4.5                4.5               cbl2   

                             beer_style  review_palate  review_taste  \
0                            Hefeweizen            1.5           1.5   
1                    English Strong Ale            3.0           3.0   
2                Foreign / Export Stout            3.0           3.0   
3                       German Pilsener            2.5           3.0   
4        American Double / Imperial IPA            4.0           4.5   
...                                 ...            ...           ...   
1586609                     Pumpkin Ale            4.0           4.0   
1586610                     Pumpkin Ale            2.0           4.0   
1586611                     Pumpkin Ale            3.5           4.0   
1586612                     Pumpkin Ale            4.5           4.5   
1586613                     Pumpkin Ale            4.5           4.5   

                      beer_name  beer_abv  beer_beerid  
0                  Sausa Weizen       5.0        47986  
1                      Red Moon       6.2        48213  
2        Black Horse Black Beer       6.5        48215  
3                    Sausa Pils       5.0        47969  
4                 Cauldron DIPA       7.7        64883  
...                         ...       ...          ...  
1586609      The Horseman's Ale       5.2        33061  
1586610      The Horseman's Ale       5.2        33061  
1586611      The Horseman's Ale       5.2        33061  
1586612      The Horseman's Ale       5.2        33061  
1586613      The Horseman's Ale       5.2        33061  

[1586614 rows x 13 columns]

Note: from the above we can see that the "dirty" dataset has 1586614 reviews.

One issue with the data as it currently stands is that some reviews are of the same beer, but the beer is inconsistently labeled across reviews (e.g. the beer_abv is different, or the beer_style is different). We remove from the dataset all reviews of beers which occur as inexact (see below) duplicates. There aren't too many of these, and it is hard to tell which information is accurate; though, in theory, for collaborative methods it doesn't matter which information is accurate as long as we are able to identify that they are the same. Therefore, we could also keep reviews of inexact duplicate beers as long as we change inexact duplicate beer_beerid's to match. If we do content-based methods, then we also would need to make sure the rest of their data (beer_abv, beer_style) match (put differently, we'd need to choose a beer_abv and a beer_style to represent this beer). In this case we would probably also want to identify duplicate breweries (breweries which have different brewery_id's but only because of minor spelling differences e.g. BJ’s Restaurant & Brewhouse vs BJ’s Restaurant And Brewhouse). Note that such duplicates could also still live in our dataset even after removing all reviews of inexact duplicates, but these types of duplicates are harder to identify.

EXACT duplicate beers are beers that agree on beer_beerid, beer_name, beer_style, beer_abv, brewery_name, and brewery_id.

INEXACT duplicate beers are beers that agree on beer_name and brewery_name (but differ on some other data and hence have different beer_beerid's).

In [4]:
clean_reviews = clean.remove_dup_beer_rows(clean.remove_null_rows(reviews))
clean_reviews

brewery_id                 brewery_name         review_time  \
0             10325              Vecchio Birraio 2009-02-16 20:57:03   
1             10325              Vecchio Birraio 2009-03-01 13:44:57   
2             10325              Vecchio Birraio 2009-03-01 14:10:04   
3             10325              Vecchio Birraio 2009-02-15 19:12:25   
4              1075      Caldera Brewing Company 2010-12-30 18:53:26   
...             ...                          ...                 ...   
1586609       14359  The Defiant Brewing Company 2006-11-05 00:01:32   
1586610       14359  The Defiant Brewing Company 2006-10-17 01:29:26   
1586611       14359  The Defiant Brewing Company 2006-10-13 01:21:53   
1586612       14359  The Defiant Brewing Company 2006-10-05 04:37:24   
1586613       14359  The Defiant Brewing Company 2006-10-04 23:41:59   

         review_overall  review_aroma  review_appearance review_profilename  \
0                   1.5           2.0                2.5            stcules   
1                   3.0           2.5                3.0            stcules   
2                   3.0           2.5                3.0            stcules   
3                   3.0           3.0                3.5            stcules   
4                   4.0           4.5                4.0     johnmichaelsen   
...                 ...           ...                ...                ...   
1586609             5.0           4.0                3.5         maddogruss   
1586610             4.0           5.0                2.5          yelterdow   
1586611             4.5           3.5                3.0           TongoRad   
1586612             4.0           4.5                4.5           dherling   
1586613             5.0           4.5                4.5               cbl2   

                             beer_style  review_palate  review_taste  \
0                            Hefeweizen            1.5           1.5   
1                    English Strong Ale            3.0           3.0   
2                Foreign / Export Stout            3.0           3.0   
3                       German Pilsener            2.5           3.0   
4        American Double / Imperial IPA            4.0           4.5   
...                                 ...            ...           ...   
1586609                     Pumpkin Ale            4.0           4.0   
1586610                     Pumpkin Ale            2.0           4.0   
1586611                     Pumpkin Ale            3.5           4.0   
1586612                     Pumpkin Ale            4.5           4.5   
1586613                     Pumpkin Ale            4.5           4.5   

                      beer_name  beer_abv  beer_beerid  
0                  Sausa Weizen       5.0        47986  
1                      Red Moon       6.2        48213  
2        Black Horse Black Beer       6.5        48215  
3                    Sausa Pils       5.0        47969  
4                 Cauldron DIPA       7.7        64883  
...                         ...       ...          ...  
1586609      The Horseman's Ale       5.2        33061  
1586610      The Horseman's Ale       5.2        33061  
1586611      The Horseman's Ale       5.2        33061  
1586612      The Horseman's Ale       5.2        33061  
1586613      The Horseman's Ale       5.2        33061  

[1581515 rows x 13 columns]

After removing all reviews of inexact duplicate beers, we still have 1581515 reviews (~99.68% of the original data).

As can be seen below, any two beers in clean_reviews that share the same beer_name and brewery_name share all the same information.

In [5]:
beers = clean.get_beers(clean_reviews) # gets all beers represented, only nixing extranious copies of EXACT duplicates.
beers[beers.duplicated(['beer_name', 'brewery_name'])] # find any inexact duplicates

Empty DataFrame
Columns: [brewery_id, brewery_name, beer_style, beer_name, beer_abv, beer_beerid]
Index: []